File: libs/caramba-core/api/route_report.go
package api

import (
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/semanticparadox/caramba/libs/caramba-core/engine"
	"github.com/semanticparadox/caramba/libs/caramba-core/routing"
)

// Отчёт о применённой маршрутизации.
//
// Владелец сказал про блок рекламы и стриминг: «непонятно работают или нет».
// Это не претензия к экрану — до этого файла в приложении не было НИ ОДНОГО
// канала, по которому такой ответ мог бы прийти. ApplyPreset сообщает только
// «идентификатор существует»; дальше пресет молча теряет части себя:
//
//   - routing.BuildWith выбрасывает недоступный rule-provider ВМЕСТЕ с
//     правилами, которые на него ссылаются, и «Россия (умный)» без ru-blocked
//     снаружи неотличим от «Россия (умный)» с ним;
//   - пресеты adblock и streaming это ЧИСТЫЕ теги GEOSITE, а секция geox-url
//     пишется только при Bootstrap.Managed (profile.applyGeoX), то есть лишь
//     под доверенным каталогом с подписанными хешами. В остальных сборках ядро
//     остаётся на зашитом умолчании mihomo — четырёх ссылках на
//     github.com/MetaCubeX/meta-rules-dat, недоступных ровно там, где этот
//     клиент и нужен. Тег без базы не значит ничего, и это НЕ видно;
//   - relay на raw-пути отбрасывается, а на панельном уезжает в панель, чья
//     цепочка в теле CLASH не выражается вовсе.
//
// Отчёт снимается на подъёме и отвечает на вопрос «что применено», а не «что
// выбрано». Там, где ядро не знает, здесь стоит «неизвестно» с причиной: это
// законное значение, и придумывать вместо него здоровый ответ запрещено.

// Значения RouteReport.Source.
const (
	// RouteSourcePreset: маршрутизация собрана из встроенного пресета.
	RouteSourcePreset = "preset"
	// RouteSourceCustom: конфигурация подана напрямую (SetRouting) — пресета
	// за ней нет, и разбирать её на источники здесь не по чему.
	RouteSourceCustom = "custom"
	// RouteSourceCoreDefault: ни пресета, ни своей конфигурации. Правила
	// выбирает profile по умолчанию, и их состав ядру api неизвестен.
	RouteSourceCoreDefault = "core_default"
)

// Причина, по которой отчёта нет вообще.
const (
	// RouteUnknownNotRaised: туннель ни разу не поднимался в этом экземпляре
	// ядра. Отчёт снимается на подъёме, и до него сказать нечего.
	RouteUnknownNotRaised = "not_raised"
)

// Состояние базы GeoSite.dat, от которой зависят ВСЕ правила GEOSITE.
const (
	// GeositeNotRequired: во вошедших правилах нет ни одного матчера GEOSITE.
	GeositeNotRequired = "not_required"
	// GeositeVerified: каталог назвал базу, её байты сошлись с подписанным
	// sha256 и лежат на диске. Это единственное состояние, про которое ядро
	// говорит «да».
	GeositeVerified = "verified"
	// GeositePresent: файл на диске есть, но подписи под ним нет — ядро
	// скачало его само по зашитому умолчанию. Правила скорее всего работают,
	// но происхождение базы клиент не проверял.
	GeositePresent = "present"
	// GeositeRefused: каталог доверен, но базу он не назвал, поэтому geox-url
	// получил ПУСТОЙ адрес и ядро откажется её качать (инвариант 12: отказ, а
	// не откат на GitHub). Все правила GEOSITE мертвы, и это точно известно.
	GeositeRefused = "refused"
	// GeositeUnknown: файла нет и подписанного адреса нет. Скачает ли ядро
	// базу с зашитого умолчания, отсюда не видно.
	GeositeUnknown = "unknown"
)

// Машинные коды причин по базе GEOSITE.
const (
	// GeositeReasonUnmanaged: Bootstrap.Managed снят, секция geox-url в конфиг
	// не пишется (profile.applyGeoX), ядро идёт на meta-rules-dat.
	GeositeReasonUnmanaged = "geox_unmanaged"
	// GeositeReasonNotInCatalog: каталог доверен, но GeoSite.dat в нём нет.
	GeositeReasonNotInCatalog = "not_in_catalog"
	// GeositeReasonFileMissing: адрес подписан, а файла на диске нет.
	GeositeReasonFileMissing = "file_missing"
)

// Состояние запрошенной страны входа.
const (
	// RouteRelayNotRequested: вход не выбирали.
	RouteRelayNotRequested = "not_requested"
	// RouteRelayIgnored: выбор был, и этот путь его применить не смог. Причина
	// лежит в Capability тем же кодом, каким её отдаёт Capabilities.
	RouteRelayIgnored = "ignored"
	// RouteRelaySent: страна ушла в панель параметром relay_country. Построила
	// ли панель цепочку — отдельный вопрос, см. DialerProxySeen.
	RouteRelaySent = "sent"
)

// GeositeReport — доступность базы, без которой правила GEOSITE это ничто.
type GeositeReport struct {
	// Required: во вошедших правилах есть хотя бы один матчер GEOSITE.
	Required bool `json:"required"`
	// Tags — те самые теги, ради которых база нужна.
	Tags []string `json:"tags,omitempty"`
	// State — одно из Geosite* выше.
	State string `json:"state"`
	// Reason — машинный код (GeositeReason*). Пусто при verified и
	// not_required.
	Reason string `json:"reason,omitempty"`
	// Detail — пояснение на английском для журнала, не для пользователя.
	Detail string `json:"detail,omitempty"`
	// Path — файл, который ядро проверяло. Отдаётся всегда, даже когда его
	// там нет: «искал вот здесь» это часть ответа.
	Path string `json:"path"`
	// SizeBytes — размер найденного файла, 0 если файла нет.
	SizeBytes int64 `json:"size_bytes"`
	// URL — подписанный адрес базы. Пусто означает, что подписанного адреса
	// нет, а не что адрес есть и он пустой.
	URL string `json:"url,omitempty"`
}

// RouteRelayReport — что стало с выбранной страной входа.
type RouteRelayReport struct {
	// Requested — то, что выбрал пользователь (ISO-2 либо имя). Пусто, когда
	// выбора не было.
	Requested string `json:"requested,omitempty"`
	// State — одно из RouteRelay* выше.
	State string `json:"state"`
	// Capability — та же запись, что отдаёт Capabilities, когда выбор
	// отброшен. Обвязка берёт текст по её Reason и не заводит вторую
	// формулировку того же факта.
	Capability *Capability `json:"capability,omitempty"`
	// DialerProxySeen: в теле конфига, из которого поднялся туннель, найден
	// ключ dialer-proxy — единственная форма, которой mihomo выражает цепочку
	// вход→выход.
	//
	// Это НАБЛЮДЕНИЕ над применённым телом, а не убеждение о панели. Сегодня
	// панель отдаёт ядру формат CLASH, а генератор CLASH цепочку не строит:
	// поле останется false даже при принятом relay_country. Если панель
	// когда-нибудь начнёт её отдавать, поле станет true само, без правки здесь.
	DialerProxySeen bool `json:"dialer_proxy_seen"`
	// Detail — пояснение на английском для журнала.
	Detail string `json:"detail,omitempty"`
}

// RouteReport — что подъём РЕАЛЬНО применил к маршрутизации.
type RouteReport struct {
	// Known ложно ровно в одном случае: подъёма ещё не было. Всё остальное,
	// включая «ядро решает само», это известное состояние с именем.
	Known bool `json:"known"`
	// Reason — машинный код (RouteUnknown*), когда Known ложно.
	Reason string `json:"reason,omitempty"`
	// Detail — пояснение на английском для журнала.
	Detail string `json:"detail,omitempty"`
	// RaisedAtMs — момент подъёма, unix-миллисекунды. 0 при Known=false.
	RaisedAtMs int64 `json:"raised_at_ms,omitempty"`
	// TunnelUp — поднят ли туннель ПРЯМО СЕЙЧАС. Отчёт переживает Down
	// намеренно: вопрос «почему не резалась реклама» задают уже отключившись.
	TunnelUp bool `json:"tunnel_up"`
	// Source — одно из RouteSource* выше.
	Source string `json:"source,omitempty"`
	// Preset — отчёт сборки пресета. nil, когда Source не preset.
	Preset *routing.BuildReport `json:"preset,omitempty"`
	// Rules — сколько правил применено. nil означает «неизвестно»: при
	// RouteSourceCoreDefault состав правил выбирает profile, и этот слой их не
	// считает. Ноль и «не знаю» здесь разные ответы.
	Rules *int `json:"rules"`
	// Geosite — доступность базы GEOSITE.
	Geosite GeositeReport `json:"geosite"`
	// Relay — судьба выбранной страны входа.
	Relay RouteRelayReport `json:"relay"`
	// Ignored — полный список отброшенных этим подъёмом выборов, тот же, что
	// в UpResult.Ignored.
	Ignored []Capability `json:"ignored,omitempty"`
}

// routeSnapshot — сырьё отчёта, снятое внутри Up.
//
// Отдельный тип, потому что часть полей известна ДО старта движка (план
// загрузки, сборка пресета), а часть только после (момент подъёма). Собирать
// отчёт из полей Core на чтении нельзя: к тому времени presetID и relayCountry
// уже могли смениться, и отчёт рассказывал бы про выбор, которого подъём не
// видел.
type routeSnapshot struct {
	source string
	preset *routing.BuildReport
	rules  *int
	// geositeTags — теги GEOSITE во ВОШЕДШИХ правилах, без повторов.
	//
	// Считается по применённой конфигурации, а не по пресету: правила могли
	// прийти и напрямую (SetRouting). Пустой список означает, что база
	// GeoSite.dat этому подъёму не нужна вовсе, и это тоже ответ.
	geositeTags []string
	geoManaged  bool
	geoURL      string
	relay       RouteRelayReport
	ignored     []Capability
	raisedAt    time.Time
}

// buildRouteReport достраивает снимок до отчёта: проверяет базу GEOSITE на
// диске и подставляет текущее состояние движка.
func (c *Core) buildRouteReport(s routeSnapshot, tunnelUp bool) RouteReport {
	rep := RouteReport{
		Known:      true,
		RaisedAtMs: s.raisedAt.UnixMilli(),
		TunnelUp:   tunnelUp,
		Source:     s.source,
		Preset:     s.preset,
		Rules:      s.rules,
		Relay:      s.relay,
		Ignored:    s.ignored,
	}

	g := GeositeReport{
		Path: filepath.Join(c.workDir, geoSiteFile),
		URL:  s.geoURL,
		Tags: s.geositeTags,
	}
	g.Required = len(g.Tags) > 0

	st, err := os.Stat(g.Path)
	onDisk := err == nil && !st.IsDir()
	if onDisk {
		g.SizeBytes = st.Size()
	}

	switch {
	case !g.Required:
		// Без пресета profile.applyRules эмитит только GEOIP,private и MATCH:
		// страновые geo-правила туда намеренно не зашиты, а собственные
		// правила подписки перезаписываются целиком. Матчера GEOSITE в
		// применённых правилах нет, и база не нужна.
		g.State = GeositeNotRequired
		g.Detail = "no GEOSITE matcher survived into the applied rules, so the database is never consulted"
	case s.geoManaged && g.URL == "":
		// Инвариант 12: при доверенном каталоге неназванный ресурс получает
		// пустой адрес, и ядро отказывается его качать, а не идёт на GitHub.
		g.State, g.Reason = GeositeRefused, GeositeReasonNotInCatalog
		g.Detail = "the trusted catalog is in force but does not name GeoSite.dat, so geox-url carries an empty address and the engine refuses to download it: every GEOSITE rule is dead"
	case s.geoManaged && onDisk:
		g.State = GeositeVerified
		g.Detail = "the catalog named GeoSite.dat and its bytes matched the signed sha256 before they were written here"
	case s.geoManaged:
		g.State, g.Reason = GeositeUnknown, GeositeReasonFileMissing
		g.Detail = "the catalog names GeoSite.dat but no file is present at the path the engine reads"
	case onDisk:
		g.State, g.Reason = GeositePresent, GeositeReasonUnmanaged
		g.Detail = "a GeoSite.dat is present but unsigned: geox-url is not written without a trusted catalog, so the engine fetched it from its compiled default"
	default:
		g.State, g.Reason = GeositeUnknown, GeositeReasonUnmanaged
		g.Detail = "no trusted catalog and no local GeoSite.dat: geox-url is not written, and whether the engine can reach its compiled default (github.com/MetaCubeX/meta-rules-dat) is not observable from here"
	}
	rep.Geosite = g
	return rep
}

// newRouteRelayReport описывает судьбу выбранной страны входа.
//
// rawYAML это тело, из которого собран конфиг. Ключ dialer-proxy ищется в нём
// подстрокой намеренно: это ровно то имя, которым mihomo выражает цепочку, и
// наблюдение над применённым телом честнее, чем утверждение о том, что панель
// умеет.
func newRouteRelayReport(requested string, ignored []Capability, rawYAML []byte) RouteRelayReport {
	out := RouteRelayReport{
		Requested:       strings.TrimSpace(requested),
		DialerProxySeen: strings.Contains(string(rawYAML), "dialer-proxy"),
	}
	if out.Requested == "" {
		out.State = RouteRelayNotRequested
		return out
	}
	for i := range ignored {
		if ignored[i].Name == CapNameRelayChaining {
			cap := ignored[i]
			out.State, out.Capability = RouteRelayIgnored, &cap
			out.Detail = "the entry country was dropped by this path; see capability.reason"
			return out
		}
	}
	out.State = RouteRelaySent
	if out.DialerProxySeen {
		out.Detail = "relay_country was sent to the panel and the applied body carries a dialer-proxy chain"
		return out
	}
	out.Detail = "relay_country was sent to the panel, but the applied body carries no dialer-proxy key, which is the only way mihomo expresses an entry-exit chain: the core cannot confirm a chain was built"
	return out
}

// RouteReportJSON отдаёт отчёт о маршрутизации последнего подъёма.
//
// Читающий вызов: он ничего не запрашивает по сети и ничего не применяет.
func (c *Core) RouteReportJSON() (string, error) {
	c.mu.Lock()
	snap := c.lastRoute
	c.mu.Unlock()
	if snap == nil {
		return toJSONString(RouteReport{
			Reason: RouteUnknownNotRaised,
			Detail: "no tunnel has been raised by this core instance, so there is nothing applied to report on",
			Geosite: GeositeReport{
				State:  GeositeUnknown,
				Reason: RouteUnknownNotRaised,
				Path:   filepath.Join(c.workDir, geoSiteFile),
			},
			Relay: RouteRelayReport{State: RouteRelayNotRequested},
		})
	}
	st, _ := c.engine.Status()
	return toJSONString(c.buildRouteReport(*snap, st.State == engine.StateConnected))
}

File: apps/caramba-client/lib/features/settings/route_report.dart
/// Разбор отчёта ядра о ФАКТИЧЕСКИ применённом маршруте.
///
/// Владелец сформулировал претензию так: «настройки по типу блок рекламы или
/// стриминг непонятно работают или нет». До появления `Core.RouteReportJSON()`
/// ответить было нечем — приложение отдавало ядру идентификатор пресета и
/// больше про него ничего не слышало, поэтому единственно возможной формой
/// ответа была галочка «включено», которая означала лишь «мы попросили».
///
/// Здесь эта галочка разбирается на два разных утверждения:
///   * пресет ПРИМЕНЁН — ядро подняло туннель именно с ним;
///   * источники его правил РАЗРЕШИЛИСЬ (или не разрешились, и тогда названо,
///     какой именно список не доехал и почему).
///
/// Ни одно поле здесь не додумывается. `known: false` это «подъёма не было», а
/// не «всё хорошо»; `rules: null` это «состав правил выбрало ядро», а не ноль;
/// `geosite.state: unknown` это «проверить нечем», а не «работает». Текст
/// `detail` ядра написан по-английски для журнала и пользователю не
/// показывается — на экран идут только переводы машинных кодов.
library;

import 'dart:convert';

import 'package:flutter_riverpod/flutter_riverpod.dart';

import 'package:caramba_client/domain/offering/route_presets.dart';
import 'package:caramba_client/state/providers.dart';
import 'package:caramba_client/state/vpn_state.dart';

/// Откуда ядро взяло состав правил.
enum AppliedRouteSource {
  /// Пресет из реестра ядра.
  preset,

  /// Правила пришли готовыми (`SetRouting`).
  custom,

  /// Ядро выбрало правила само: `GEOIP,private` + `MATCH`.
  coreDefault,

  /// Поля `source` в отчёте нет.
  unknown,
}

/// Судьба одного внешнего списка правил (`preset.sources[]`).
enum RuleSourceState {
  /// Список взят из проверенного локального файла.
  file,

  /// Список выпущен ссылкой на зеркало: доедет ли он, отсюда не видно.
  mirror,

  /// Список выброшен из сборки: его правил в туннеле нет.
  dropped,

  /// Незнакомое ядру состояние.
  unknown,
}

/// Состояние базы GEOSITE — той самой, без которой блок рекламы и стриминг
/// это пустые правила.
enum GeositeState {
  /// Пресету GEOSITE не нужен.
  notRequired,

  /// База проверена каталогом.
  verified,

  /// Файл на месте, но подписи под ним нет.
  present,

  /// Каталог в силе и `GeoSite.dat` в нём не назван: `geox-url` пуст, движок
  /// базу не скачает, и КАЖДОЕ правило GEOSITE заведомо мертво.
  refused,

  /// Проверить нечем.
  unknown,
}

/// Что стало со страной входа.
enum AppliedRelayState { notRequested, ignored, sent, unknown }

/// Один внешний список правил и его судьба.
class AppliedRuleSource {
  final String name;
  final RuleSourceState state;

  /// Машинный код ядра: `not_in_catalog`, `no_mirror`.
  final String? reason;

  /// Сколько правил давал бы этот список.
  final int rules;

  /// Сколько из них доехало до сборки.
  final int keptRules;

  const AppliedRuleSource({
    required this.name,
    required this.state,
    this.reason,
    this.rules = 0,
    this.keptRules = 0,
  });

  bool get isDropped => state == RuleSourceState.dropped;

  /// Русский текст судьбы списка. Английский `detail` ядра сюда не попадает
  /// намеренно: он для журнала.
  String get message => switch (state) {
    RuleSourceState.file => 'список на месте, правил: $keptRules',
    RuleSourceState.mirror =>
      'выпущен ссылкой на зеркало оператора; доехал ли он, ядро не видит',
    RuleSourceState.dropped => switch (reason) {
      'no_mirror' =>
        'выброшен: файла нет и адреса зеркала тоже, скачивать неоткуда '
            '(потеряно правил: $rules)',
      'not_in_catalog' =>
        'выброшен: каталог оператора этот список не подписал '
            '(потеряно правил: $rules)',
      _ => 'выброшен из сборки (потеряно правил: $rules)',
    },
    RuleSourceState.unknown => 'состояние списка ядру неизвестно',
  };

  static AppliedRuleSource? _parse(Object? raw) {
    if (raw is! Map) return null;
    final m = raw.cast<String, Object?>();
    final name = _str(m['name']);
    if (name.isEmpty) return null;
    return AppliedRuleSource(
      name: name,
      state: switch (_str(m['state'])) {
        'file' => RuleSourceState.file,
        'mirror' => RuleSourceState.mirror,
        'dropped' => RuleSourceState.dropped,
        _ => RuleSourceState.unknown,
      },
      reason: _strOrNull(m['reason']),
      rules: _int(m['rules']) ?? 0,
      keptRules: _int(m['kept_rules']) ?? 0,
    );
  }
}

/// Применённый пресет: то, что ядро действительно собрало.
class AppliedPreset {
  final String id;
  final String name;
  final String emoji;
  final String countryCode;

  /// Сколько правил вошло в сборку.
  final int rules;

  /// Сколько правил потеряно вместе с не доехавшими списками.
  final int droppedRules;

  /// Теги GEOSITE, на которых держится пресет.
  final List<String> geositeTags;

  final List<AppliedRuleSource> sources;

  const AppliedPreset({
    required this.id,
    required this.name,
    required this.emoji,
    required this.countryCode,
    required this.rules,
    required this.droppedRules,
    required this.geositeTags,
    required this.sources,
  });

  /// Списки, которых в сборке нет.
  List<AppliedRuleSource> get droppedSources =>
      sources.where((s) => s.isDropped).toList(growable: false);

  /// Режет ли этот пресет рекламу — по реестру ядра, а не по имени.
  /// `null` — пресета с таким id в зеркале реестра нет (ядро ушло вперёд).
  bool? get blocksAds => coreRoutePresetById(id)?.blocksAds;

  /// Уводит ли он стриминг в туннель.
  bool? get routesStreaming => coreRoutePresetById(id)?.routesStreaming;

  static AppliedPreset? _parse(Object? raw) {
    if (raw is! Map) return null;
    final m = raw.cast<String, Object?>();
    final id = _str(m['preset_id']);
    if (id.isEmpty) return null;
    return AppliedPreset(
      id: id,
      name: _str(m['preset_name']),
      emoji: _str(m['emoji']),
      countryCode: _str(m['country']),
      rules: _int(m['rules']) ?? 0,
      droppedRules: _int(m['dropped_rules']) ?? 0,
      geositeTags: _strList(m['geosite_tags']),
      sources: <AppliedRuleSource>[
        if (m['sources'] is List)
          for (final e in m['sources']! as List)
            if (AppliedRuleSource._parse(e) case final s?) s,
      ],
    );
  }
}

/// Состояние базы GEOSITE в применённой сборке.
class AppliedGeosite {
  /// Нужна ли база этому пресету вообще.
  final bool required;

  final GeositeState state;

  /// Машинный код: `geox_unmanaged`, `not_in_catalog`, `file_missing`,
  /// `not_raised`.
  final String? reason;

  final List<String> tags;

  const AppliedGeosite({
    required this.required,
    required this.state,
    this.reason,
    this.tags = const <String>[],
  });

  static const AppliedGeosite nothing = AppliedGeosite(
    required: false,
    state: GeositeState.unknown,
    reason: 'not_raised',
  );

  /// Работают ли правила GEOSITE. `true` — база проверена, `false` — она
  /// заведомо недоступна, `null` — проверить нечем.
  bool? get resolved => switch (state) {
    GeositeState.verified || GeositeState.present => true,
    GeositeState.refused => false,
    GeositeState.notRequired || GeositeState.unknown => null,
  };

  String get message => switch (state) {
    GeositeState.notRequired => 'Базе GEOSITE здесь нечего делать: правила '
        'пресета её не спрашивают.',
    GeositeState.verified => 'База GEOSITE проверена каталогом оператора.',
    GeositeState.present =>
      'Файл базы GEOSITE на месте, но подписи оператора под ним нет.',
    GeositeState.refused =>
      'Каталог оператора не назвал GeoSite.dat, адрес загрузки пуст — все '
          'правила GEOSITE в этой сборке мертвы.',
    GeositeState.unknown => switch (reason) {
      'geox_unmanaged' =>
        'Подтвердить базу GEOSITE нечем: доверенного каталога нет, и достанет '
            'ли движок свою встроенную базу — отсюда не видно.',
      'file_missing' => 'Локального файла базы GEOSITE нет.',
      'not_in_catalog' => 'Каталог оператора базу GEOSITE не называет.',
      'not_raised' => 'Туннель ещё не поднимали.',
      _ => 'Состояние базы GEOSITE ядру неизвестно.',
    },
  };

  static AppliedGeosite _parse(Object? raw) {
    if (raw is! Map) return nothing;
    final m = raw.cast<String, Object?>();
    return AppliedGeosite(
      required: m['required'] == true,
      state: switch (_str(m['state'])) {
        'not_required' => GeositeState.notRequired,
        'verified' => GeositeState.verified,
        'present' => GeositeState.present,
        'refused' => GeositeState.refused,
        _ => GeositeState.unknown,
      },
      reason: _strOrNull(m['reason']),
      tags: _strList(m['tags']),
    );
  }
}

/// Судьба страны входа в применённой сборке.
class AppliedRelay {
  /// Что просило приложение (ISO-2); пусто — не просило ничего.
  final String requested;

  final AppliedRelayState state;

  /// Нашёлся ли в применённом теле ключ `dialer-proxy` — единственный, которым
  /// mihomo строит цепочку. Это НАБЛЮДЕНИЕ над телом, а не обещание панели.
  final bool dialerProxySeen;

  /// Машинный код возможности (`raw_profile`, ...); `null` — ядро молчит.
  final String? capabilityReason;

  const AppliedRelay({
    required this.requested,
    required this.state,
    required this.dialerProxySeen,
    this.capabilityReason,
  });

  static const AppliedRelay nothing = AppliedRelay(
    requested: '',
    state: AppliedRelayState.notRequested,
    dialerProxySeen: false,
  );

  /// Цепочка подтверждена телом конфига.
  bool get chained => dialerProxySeen;

  String get message => switch (state) {
    AppliedRelayState.notRequested => 'Вход не запрашивался.',
    AppliedRelayState.ignored =>
      'Вход «$requested» этот путь выразить не может, и он отброшен.',
    AppliedRelayState.sent => dialerProxySeen
        ? 'Вход «$requested» запрошен, и цепочка в применённом конфиге есть.'
        : 'Вход «$requested» ушёл оператору, но цепочки в применённом конфиге '
              'нет: трафик идёт прямо на выход.',
    AppliedRelayState.unknown => 'Состояние входа ядру неизвестно.',
  };

  static AppliedRelay _parse(Object? raw) {
    if (raw is! Map) return nothing;
    final m = raw.cast<String, Object?>();
    final cap = m['capability'];
    return AppliedRelay(
      requested: _str(m['requested']),
      state: switch (_str(m['state'])) {
        'not_requested' => AppliedRelayState.notRequested,
        'ignored' => AppliedRelayState.ignored,
        'sent' => AppliedRelayState.sent,
        _ => AppliedRelayState.unknown,
      },
      // Молчание про ключ не значит, что он есть.
      dialerProxySeen: m['dialer_proxy_seen'] == true,
      capabilityReason: cap is Map
          ? _strOrNull(cap.cast<String, Object?>()['reason'])
          : null,
    );
  }
}

/// Полный отчёт ядра о применённом маршруте.
class AppliedRoute {
  /// Было ли вообще что применять. `false` — туннель этим экземпляром ядра ещё
  /// не поднимали, и это НЕ «всё хорошо».
  final bool known;

  /// Машинный код, почему неизвестно (`not_raised`).
  final String? reason;

  /// Держится ли туннель прямо сейчас.
  final bool tunnelUp;

  final AppliedRouteSource source;

  /// Применённый пресет; `null` — правила выбрало ядро или пришли готовыми.
  final AppliedPreset? preset;

  /// Сколько правил в сборке. `null` — состав выбрало ядро, и это ответ,
  /// отличный от нуля.
  final int? rules;

  final AppliedGeosite geosite;
  final AppliedRelay relay;

  /// Ядро этой сборки отчёт не отдаёт вовсе (пустая строка на проводе).
  final bool supported;

  const AppliedRoute({
    required this.known,
    required this.tunnelUp,
    required this.source,
    required this.geosite,
    required this.relay,
    this.reason,
    this.preset,
    this.rules,
    this.supported = true,
  });

  /// Ядро отчёта не знает: старая сборка или мост недоступен.
  static const AppliedRoute unsupported = AppliedRoute(
    known: false,
    tunnelUp: false,
    source: AppliedRouteSource.unknown,
    geosite: AppliedGeosite.nothing,
    relay: AppliedRelay.nothing,
    reason: 'no_bridge',
    supported: false,
  );

  /// Есть ли что показывать пользователю.
  bool get hasReport => supported && known;

  /// Режет ли применённая сборка рекламу. `null` — сборка про это молчит
  /// (пресета нет или он не из зеркала реестра).
  bool? get blocksAds => preset?.blocksAds;

  bool? get routesStreaming => preset?.routesStreaming;

  /// Правила, зависящие от GEOSITE, точно мертвы.
  bool get geositeDead => geosite.required && geosite.resolved == false;

  /// Списки, не доехавшие до сборки.
  List<AppliedRuleSource> get droppedSources =>
      preset?.droppedSources ?? const <AppliedRuleSource>[];

  /// Разбирает отчёт ядра. Пустая строка или мусор — [unsupported]: выдумывать
  /// здоровый отчёт нельзя, иначе экран покажет зелёную галочку там, где ядро
  /// не сказало ничего.
  factory AppliedRoute.parse(String raw) {
    if (raw.trim().isEmpty) return unsupported;
    final Object? decoded;
    try {
      decoded = jsonDecode(raw);
    } catch (_) {
      return unsupported;
    }
    if (decoded is! Map) return unsupported;
    final m = decoded.cast<String, Object?>();
    return AppliedRoute(
      known: m['known'] == true,
      reason: _strOrNull(m['reason']),
      tunnelUp: m['tunnel_up'] == true,
      source: switch (_str(m['source'])) {
        'preset' => AppliedRouteSource.preset,
        'custom' => AppliedRouteSource.custom,
        'core_default' => AppliedRouteSource.coreDefault,
        _ => AppliedRouteSource.unknown,
      },
      preset: AppliedPreset._parse(m['preset']),
      // Именно `_int`, а не `?? 0`: ноль правил и «состав выбрало ядро» —
      // разные ответы, и склеивать их значит терять единственный, ради
      // которого отчёт и появился.
      rules: _int(m['rules']),
      geosite: AppliedGeosite._parse(m['geosite']),
      relay: AppliedRelay._parse(m['relay']),
    );
  }
}

/// Отчёт ядра о применённом маршруте, перечитываемый на каждой смене стадии
/// туннеля.
///
/// Стадия здесь не ради значения, а ради зависимости: отчёт меняется ровно в
/// момент подъёма, и без подписки карточка показывала бы вечное «подъёма не
/// было» уже после подключения.
final appliedRouteProvider = FutureProvider<AppliedRoute>((ref) async {
  ref.watch(vpnProvider.select((s) => s.stage));
  try {
    return AppliedRoute.parse(await ref.watch(vpnConnectionProvider).routeReport());
  } catch (_) {
    // Мост может отсутствовать в этой сборке. Это «нечем проверить», а не
    // ошибка пользователя, и наверх идёт то же состояние, что и у пустой
    // строки.
    return AppliedRoute.unsupported;
  }
});

String _str(Object? v) => v is String ? v.trim() : '';

String? _strOrNull(Object? v) {
  final s = _str(v);
  return s.isEmpty ? null : s;
}

int? _int(Object? v) => v is num ? v.toInt() : null;

List<String> _strList(Object? v) => v is List
    ? <String>[
        for (final e in v)
          if (_str(e).isNotEmpty) _str(e),
      ]
    : const <String>[];
